src/librustc_typeck/check/coercion.rs

```diff
@@ -698,6 +700,32 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         })
     }
 
+    /// Same as `try_coerce()`, but without side-effects. Intended to
+    /// be used for error reporting only.
+    pub fn can_coerce(&self,
+                      expr_ty: Ty<'tcx>,
+                      target: Ty<'tcx>)
+                      -> bool {
+        // FIXME: This is a hack, but coercion wasn't made to be run
+        // in a probe. It leaks obligations and bounds and things out
+        // into the environment. For now we just save-and-restore the
+        // fulfillment context.
+        let saved_fulfillment_cx =
+            mem::replace(
+                &mut *self.inh.fulfillment_cx.borrow_mut(),
+                FulfillmentContext::new());
+        let source = self.resolve_type_vars_with_obligations(expr_ty);
```

Comment:
You don't want to swap out the fulfillment context but either way should be fine.